lib/services/dashd.js

```diff
@@ -856,11 +871,9 @@ Dash.prototype._checkReindex = function(node, callback) {
 Dash.prototype._loadTipFromNode = function(node, callback) {
   var self = this;
   node.client.getBestBlockHash(function(err, response) {
-    if (err && err.code === -28) {
-      log.warn(err.message);
-      return callback(self._wrapRPCError(err));
-    } else if (err) {
-      return callback(self._wrapRPCError(err));
+    if (err) {
+      callback(self._wrapRPCError(err));
```

Comment:
Is there a reason to split callback from return? It looks like should be together

Reply:
Precisely because the other way is confusing and incorrect.

Looking at the former  _code_ you would _think_ that there's a return of the callback - similar to a return of a promise or a synchronous filter function - which is incorrect.

In reality it's more of a "monkey see monkey do" thing popular among people who are unfamiliar with how JavaScript functions work  and which makes the code more difficult read and predict the behavior - as you have just discovered.

You should never return a "done" or "next" or "callback" function unless the API actually asks for it (in which case it should be named something else to avoid confusion).

When you see them separate, you can be sure that there's no strange side effects from the function being called.

When they're together (which unfortunately isn't caught by most linters even still) you always have to double check before making changes to figure out if it was accidental / negligent, or if there's some obscure purpose behind it - which is rare... but still far too common.

Reply:
I think you might be messing up functions with function results, in this case no function is not really returned back to the caller. Basically, those 2 options does not really differs from each other (because the caller is not reading any return from it), but one-line one is most commonly being used in callback style code and also it also codestyled so in the dashcore codebase

Reply:
> I think you might be messing up functions with function results.

I'm a JavaScript ninja. I know things that would make your brain bleed. I do get things mixed up now and then, but not this.

> Basically, those 2 options does not really differs from each other.

Except when they do.

> but one-line one is most commonly being used in callback style code and it also codestyled so in the dashcore codebase

Well, that's wrong and it needs to change. This is finance software. We can't be following along with people's old bad habits or using "cool" bugs in the language at the expense of readability and security.

:)

Reply:
Returning callback results it's clearly antipattern. But changing the behavior of only one function leads to inconsistency for the package. It seems to me there are two topics covered in this PR: code style and error handling. I would recommend splitting it into two PRs: error handling fix and refactoring with return callback thing for the whole package but not just one function.

Reply:
fair